apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java

```diff
@@ -195,6 +198,179 @@ public Bundle searchByLogicalId(
     return bundle;
   }
 
+  @Search
+  public Bundle searchByCoverageContract(
+      // This is very explicit as a place holder until this kind
+      // of relational search is more common.
+      @RequiredParam(name = "_has:Coverage.extension") TokenParam coverageId,
+      @OptionalParam(name = "startIndex") String startIndex,
+      RequestDetails requestDetails) {
+
+    if (coverageId.getQueryParameterQualifier() != null)
+      throw new InvalidRequestException(
+          "Unsupported query parameter qualifier: " + coverageId.getQueryParameterQualifier());
+
+    String contractCode = coverageId.getValueNotNull();
+    if (contractCode.length() != 5)
+      throw new InvalidRequestException("Unsupported query parameter value: " + contractCode);
+
+    String contractMonth =
+        coverageId.getSystem().substring(coverageId.getSystem().lastIndexOf('/') + 1);
+    CcwCodebookVariable partDContractMonth = partDCwVariableFor(contractMonth);
+    SingularAttribute<Beneficiary, String> contractMonthField = partDFieldFor(partDContractMonth);
+
+    List<String> includeIdentifiersValues = returnIncludeIdentifiersValues(requestDetails);
+    List<SetAttribute<Beneficiary, ?>> withRelations =
+        new LinkedList<SetAttribute<Beneficiary, ?>>();
+
+    if (hasHICN(includeIdentifiersValues))
+      withRelations.add((SetAttribute<Beneficiary, ?>) Beneficiary_.beneficiaryHistories);
+
+    if (hasMBI(includeIdentifiersValues))
+      withRelations.add((SetAttribute<Beneficiary, ?>) Beneficiary_.medicareBeneficiaryIdHistories);
+
+    CriteriaQuery beneficiariesQuery =
+        queryBeneficiariesBy(contractMonthField, contractCode, withRelations);
+
+    PagingArguments pagingArgs = new PagingArguments(requestDetails);
+    List<Beneficiary> matchingBeneficiaries = fetchBeneficiaries(beneficiariesQuery, pagingArgs);
+    Long count = fetchResultCount(beneficiariesQuery);
+
+    List<IBaseResource> patients =
+        matchingBeneficiaries.stream()
+            .map(
+                beneficiary -> {
+                  // Null out the unhashed HICNs if we're not supposed to be returning them
+                  if (!hasHICN(includeIdentifiersValues)) {
+                    beneficiary.setHicnUnhashed(Optional.empty());
+                  }
+                  // Null out the unhashed MBIs if we're not supposed to be returning
+                  if (!hasMBI(includeIdentifiersValues)) {
+                    beneficiary.setMedicareBeneficiaryId(Optional.empty());
+                  }
```

Comment:
@whytheplatypus We'll need the Patient records coming out of this endpoint to contain un-hashed MBIs (if requested by the `IncludeIdentifiers` param). Based on what I see in the PR, I am pretty sure this is going to be the case, but I'd like to confirm with you anyway. An important item for us to have.

Reply:
Yup this work honors the `IncludeIdentifiers` header

Reply:
👍